core/src/main/java/org/mskcc/cbio/portal/dao/DaoReferenceGenomeGene.java

```diff
@@ -0,0 +1,255 @@
+        Connection con = null;
+        PreparedStatement pstmt = null;
+        ResultSet rs = null;
+        boolean setBulkLoadAtEnd = false;
+        try {
+            MySQLbulkLoader.bulkLoadOff();
+            int rows = 0;
+            con = JdbcUtil.getDbConnection(DaoReferenceGenome.class);
+            pstmt = con.prepareStatement
+                ("UPDATE reference_genome_gene SET `CHR`=?, `CYTOBAND`=?,`EXONIC_LENGTH`=?,`START`=?, `END`=? WHERE `ENTREZ_GENE_ID`=? AND `REFERENCE_GENOME_ID`=?");
+            pstmt.setString(1, gene.getChr());
+            pstmt.setString(2, gene.getCytoband());
+            pstmt.setInt(3, gene.getExonicLength());
+            pstmt.setLong(4, gene.getStart());
+            pstmt.setLong(5, gene.getEnd());
+            pstmt.setLong(6, gene.getEntrezGeneId());
+            pstmt.setInt(7, gene.getReferenceGenomeId());
+
+            rows += pstmt.executeUpdate();
+            if (rows != 1) {
+                ProgressMonitor.logWarning("No change for " + gene.getEntrezGeneId() + " " + gene.getReferenceGenomeId() + "? Code " + rows);
+            }
+
+            return rows;
+        } catch (SQLException e) {
+            throw new DaoException(e);
+        } finally {
+            MySQLbulkLoader.bulkLoadOn();
+            JdbcUtil.closeAll(DaoGene.class, con, pstmt, rs);
+        }
+
+    }
+
+    /**
+     *
+     * Adds a new reference genome gene Record to the Database or update the existing record.
+     *
+     * @param gene Reference Genome Gene Object.
+     * @return number of records successfully added.
+     * @throws DaoException Database Error.
+     */
+    public int addOrUpdateGene(ReferenceGenomeGene gene) throws DaoException {
+
+        Connection con = null;
+        PreparedStatement pstmt = null;
+        ResultSet rs = null;
+        try {
+            int rows = 0;
+            ReferenceGenomeGene existingGene = null;
+
+            existingGene = getGene(gene.getEntrezGeneId(), gene.getReferenceGenomeId());
+
+            if (existingGene == null) {
+                //add gene, referring to this genetic entity
+                con = JdbcUtil.getDbConnection(DaoGene.class);
+                pstmt = con.prepareStatement
+                    ("INSERT INTO `reference_genome_gene` (`ENTREZ_GENE_ID`, `REFERENCE_GENOME_ID`,`CHR`,`CYTOBAND`,`EXONIC_LENGTH`,`START`,`END`) "
+                        + "VALUES (?,?,?,?,?,?,?)");
+                pstmt.setLong(1, gene.getEntrezGeneId());
+                pstmt.setInt(2, gene.getReferenceGenomeId());
+                pstmt.setString(3, gene.getChr());
+                pstmt.setString(4, gene.getCytoband());
+                pstmt.setInt(5, gene.getExonicLength());
+                pstmt.setLong(6, gene.getStart());
+                pstmt.setLong(7, gene.getEnd());
+                rows += pstmt.executeUpdate();
+            }
+            else {
+                rows += updateGene(existingGene);
+            }
+            return rows;
```

Comment:
again, I wonder whether we really need to tell the caller whether a change happened (versus no change was necessary). Making this function void would avoid the bother, or making it type boolean and having statements like `return pstmt.executeUpdate() == 1` or `return updateGene(existingGene)` seems more clear to me than returning an integer that then needs to be interpreted by the caller.

Reply:
agree, changed to void.